src/dmd/root/file.d

```diff
@@ -22,6 +23,375 @@ import dmd.root.filename;
+                    assert(0);
+                }
+            }
+            createMapping(filename, File.size(handle));
+        }
+        else static assert(0);
+
+        // Save the name for later. Technically there's no need: on Linux one can use readlink on /proc/self/fd/NNN.
+        // On BSD and OSX one can use fcntl with F_GETPATH. On Windows one can use GetFileInformationByHandleEx.
+        // But just saving the name is simplest, fastest, and most portable...
+        import core.stdc.string : strlen;
+        name = filename[0 .. filename.strlen + 1].idup.ptr;
+    }
+
+    // Not copyable or assignable (for now).
+    @disable this(const FileMapping!Datum rhs);
+    @disable void opAssign(const ref FileMapping!Datum rhs);
+
+    /**
+    Frees resources associated with this mapping. However, it does not deallocate the name.
+    */
+    ~this() pure nothrow
+    {
+        if (!active)
+            return;
+        fakePure({
+            version (Posix)
+            {
+                import core.sys.posix.sys.mman : munmap;
+
+                if (data.ptr && munmap(cast(void*) data.ptr, data.length) != 0)
+                {
+                    fprintf(stderr, "munmap() failed for \"%s\": %s\n", filename, strerror(errno));
+                    assert(0);
+                }
+                if (handle != invalidHandle && .close(handle) != 0)
+                {
+                    fprintf(stderr, "close() failed for \"%s\": %s\n", filename, strerror(errno));
+                    assert(0);
+                }
+            }
+            else version(Windows)
+            {
+                if (data.ptr !is null && UnmapViewOfFile(cast(void*) data.ptr) == 0)
+                {
+                    fprintf(stderr, "UnmapViewOfFile() failed for \"%s\": %d\n", filename, GetLastError());
+                    assert(0);
+                }
+                if (fileMappingObject != invalidHandle && CloseHandle(fileMappingObject) == 0)
+                {
+                    fprintf(stderr, "CloseHandle() failed for mapping of \"%s\": %d\n", filename, GetLastError());
+                    assert(0);
+                }
+                if (handle != invalidHandle && CloseHandle(handle) == 0)
+                {
+                    fprintf(stderr, "CloseHandle() failed for \"%s\": %d\n", filename, GetLastError());
+                    assert(0);
+                }
+            }
+            else static assert(0);
+        });
+    }
+
+    /**
+    Returns the zero-terminated file name associated with the mapping. Can
+    be saved beyond the lifetime of `this`.
+    */
+    const(char)* filename() const pure @nogc @safe nothrow { return name; }
+
+    /**
+    Frees resources associated with this mapping. However, it does not deallocate the name.
+    Reinitializes `this` as a fresh object that can be reused.
+    */
+    void close()
+    {
+        __dtor();
```

Comment:
I really don't like the look of calling `__dtor()` in a member function.

Reply:
I had to think about it too due to my C++ bias (where there'd be complications) but in D it really is a nice idiom. What `close` needs to do is destroy the object then put it in an initialized state.